### ISSUE TITLE
Heading levels

### ISSUE THREAD

**Issue by eereeska** [thephpleague/commonmark#989]
### Description

Please, add ability to control allowed heading levels such as [CKEditor do](https://ckeditor.com/docs/ckeditor5/latest/features/headings.html#heading-levels)

### Example

May be via env config, similar to heading_permalink:

```php
'heading_permalink' => [
    'min_heading_level' => 1,
    'max_heading_level' => 6,
],
```

To:

```php
'headings' => [
    'min_level' => 2,
    'max_level' => 6,
],
```

### Did this project help you today? Did it make you happy in any way?

Definitely love it ❤️

**Comment by eereeska** [comment]
bump

**Comment by samwilson** [comment]
Would this mean that e.g. with min_level=2 a heading with one asterisk `# Head` would be rendered as `<h2>`?

A use case that I often have is that a document with headings needs to be rendered both standalone (where it's headings would be as written) and also as subsections (where the headings would all be dropped down a level). Is that the sort of thing that this would accomplish?

**Comment by colinodell** [comment]
This suggestion makes sense to me.

I'd recommend we create a separate extension for it, perhaps called `LimitHeadingsExtension` or something similar.  That extension could [observe the `DocumentParsedEvent`](https://commonmark.thephpleague.com/2.4/customization/abstract-syntax-tree/#documentparsedevent) and run some code to [adjust the heading levels](https://github.com/thephpleague/commonmark/blob/91c24291965bd6d7c46c46a12ba7492f83b1cadf/src/Extension/CommonMark/Node/Block/Heading.php#L32-L40) throughout [the parsed AST](https://commonmark.thephpleague.com/2.4/customization/abstract-syntax-tree).

Anyone interested in contributing this? 😉

**Comment by phpfui** [comment]
Looks line @eyupcanakman has implemented this in https://github.com/thephpleague/commonmark/pull/1113

The bigger problem I feel is that the [W3 HTML Validator](https://validator.w3.org/nu/?doc=http%3A%2F%2Fphpfui.com%2F) is now reporting any gap between header sizes as an error. I feel it should be a warning, but that is not my decision to make. The errors on this page are all markdown files processed with this library. You can see the Guzzle README.md showing up second in the list. The problem is it starts with H1, then goes to H3. While this may look good typographically, it is an error according to W3.org.

I am not sure how to solve this problem. Apparently you can go up any number of levels, but only down one at a time. So you upgrade lower headers, or do you downgrade higher headers?

**Comment by eyupcanakman** [comment]
@phpfui Good point, but that's a separate issue.
This PR (#1113) is strictly about clamping min/max levels to fit markdown into existing site layouts. Fixing W3C heading gaps (jumping from H1 to H3) requires a completely different approach. Mixing both in one extension would make the parsing logic unpredictable. 

W3C compliance would be better handled by a dedicated extension or a config like `warn_on_w3_violations` instead of forcing HTML changes here. 

If we open a new issue for that specific feature, I can work on integrating it.

**Comment by phpfui** [comment]
@eyupcanakman agreed it is a separate issue. Just bringing up in an existing issue.

Should I start a new issue? Not sure I have the knowledge to contribute a PR. But there needs to be an agreement on the correct solution.

**Comment by eyupcanakman** [comment]
@phpfui yeah, opening a new issue is best.

Just to clarify. I'm a contributor here, not a core maintainer. Opening an issue will let the actual maintainers decide on the correct solution. Once they agree on an approach I'd be happy to write the PR for it

**Comment by samwilson** [comment]
The above PR works in my tests. Looks good.

I do wonder if it'd be good to *shift* headings, rather than just limiting the top/bottom limits. As it is, with `'min_heading_level' => 3` for example, a document with  `## Foo\n\n### Bar` ends up with `<h3>Foo</h3><h3>Bar</h3>`, but it feels like it'd be more expected for it to have all headings moved downwards and to output `<h3>Foo</h3><h4>Bar</h4>`. Would that be something that'd be worth doing? Perhaps as a separate config option (e.g. maybe just a heading-level mapping, so everything has to be made explicit; there's only seven levels after all).

**Comment by colinodell** [comment]
This has landed! A quick summary of where things ended up:

**#1113** (merged) added the `NormalizeHeadingsExtension`, which does what this issue originally asked for:

```php
'normalize_headings' => [
    'min_level' => 2,
    'max_level' => 4,
],
```

Thanks @eyupcanakman for implementing it.

**#1127** (open) builds on that in response to the discussion here. @samwilson — you asked whether headings could be *shifted* rather than just limited, so that `## Foo` / `### Bar` with `min_level: 3` yields `<h3>` / `<h4>` rather than two `<h3>`s. That's now the default behavior: heading levels are re-derived from how deeply each one is nested, so they only collapse to the same level when the author actually wrote them as siblings. It also answers your original question — yes, with `min_level: 2` a `# Head` renders as `<h2>`.

@phpfui — that same change fixes the W3C validator complaint you raised (a document going from `<h1>` straight to `<h3>`). You were right that it deserved its own issue; it's tracked as #1115 and is fixed by the same PR.

One deliberate difference from the original example in this issue: because levels now follow nesting depth, an `###### H6` sitting under an `# H1` becomes `<h3>` rather than `<h4>`. Every heading still lands within the configured range — but the goal is a coherent outline within that range, not a strict clamp of each heading in isolation.

There's also an optional `rebase_to_min_level` for anyone who wants every document to *start* at `min_level` regardless of what the author wrote — handy for the standalone-vs-subsection case @samwilson described.

I'll close this once #1127 is merged.

**Comment by samwilson** [comment]
Thanks! This all sounds good.
